### ISSUE TITLE
build-tool-depends in Cabal regresses ghc ability to build from source distribution

### ISSUE THREAD

**Issue by mpickering** [haskell/cabal#10061]
It should be possible to build ghc from its source distribution with only `ghc`.

When the source distribution is created the .hs file is generated from the .x file. This removes the need for alex when building the project.

Recent changes in Cabal regress this by introducing `build-tool-depends: alex`.

https://gitlab.haskell.org/ghc/ghc/-/issues/24826

Potential solution: Add a flag to `Cabal-syntax` which allows the usage of `build-tool-depends` to be disabled. We can then enable this in hadrian (which ignores build-tool-depends).

One workaround we can do in hadrian is to always pass the `--with-alex` and `--with-happy` flags when configuring packages using `Cabal` because `hadrian` deals with processing these files. (Perhaps there are other workarounds?)

**Comment by mpickering** [comment]
It seems the workaround of passing `--with-happy` and `--with-alex` is not ideal because Cabal checks that the paths actually exist. So you have to be quite careful to use something which will work on all systems.

**Comment by andreabedini** [comment]
> It should be possible to build ghc from its source distribution with only ghc.

Isn't hadrian necessary to build ghc from source?

**Comment by mpickering** [comment]
@andreabedini There are bootstrap scripts to build hadrian without `cabal`. It has also previously been possible to build with just `ghc` and `cabal` (no alex or happy) so it's a regression in that sense (as reported by a user).

**Comment by andreabedini** [comment]
@mpickering
> It has also previously been possible to build with just ghc and cabal (no alex or happy) so it's a regression in that sense (as reported by a user).

`cabal` (as in `cabal-install:exe:cabal`) can fetch and build alex as needed and nothing has to come pre-installed. Can you make the intended requirements more clear?

**Comment by mpickering** [comment]
@andreabedini 

User is building from the source distribution. They have `ghc` and `cabal-install` on their system (no other build tool dependencies `alex` and `happy`). `cabal` is used to build the build-tool for `ghc` (`hadrian`). `hadrian` is used to build `ghc`. The files which are generated by `alex` and `happy` are included as part of the source distribution so there is no need for them to be available on the user's system.  Writing `build-tool-depends: alex` in this situation is inaccurate because the `.hs` files are generated already so `alex` is not a prerequisite of doing the build.

This regresses the ability to package GHC as it requires the distribution to also package `alex`.
Build dependencies (`alex`, `happy`, C toolchain etc) are provided by the distribution (not via any `cabal-install` feature). 

Does that make things more clear?

**Comment by andreabedini** [comment]
> The files which are generated by `alex` and `happy` are included as part of the source distribution so there is no need for them to be available on the user's system. Writing build-tool-depends: alex in this situation is inaccurate because the .hs files are generated already so alex is not a prerequisite of doing the build.

`Lexer.hs` is not part of Cabal-syntax's source distribution, see yourself: https://hackage.haskell.org/package/Cabal-syntax-3.12.0.0/src.

> This regresses the ability to package GHC as it requires the distribution to also package alex.
> Build dependencies (alex, happy, C toolchain etc) are provided by the distribution (not via any cabal-install feature).

Does the distribution package alex or not?

`build-tool-depends: alex` is correct because the source distribution does not ship `Lexer.hs` so it is indeed required. Either cabal-install, Hadrian or something else *has* to provide a working version of alex.

**Comment by mpickering** [comment]
@andreabedini I am talking about GHC's source distribution, for example - https://downloads.haskell.org/~ghc/9.10.1/ghc-9.10.1-src.tar.xz. In there you will find `Lexer.hs`. 

I can't comment about whether a specific distribution packages alex or not but for a long time, there has been no requirement to have alex if you want to build GHC from a source distribution.

**Comment by geekosaur** [comment]
GHC isn't alone here: there is a long history (back to the 1980s at least) of packages bundling generated files that require extra dependencies, specifically including lexer and parser generators (`alex` and `happy` for Haskell, `flex` and `bison` or `byacc` for C, etc.) to make life easier for people building the distribution and to avoid bootstrapping loops (e.g. needing a compiler to install the tools needed to install the compiler). This use case should work.

**Comment by geekosaur** [comment]
This said, I wonder if it can be done with flags, viz.
```
flag regenerate
  default: False
  manual: True
-- …
if flag(regenerate)
  build-tool-depends:
    alex,
    happy
```

**Comment by mpickering** [comment]
@geekosaur That is the solution implemented in https://github.com/haskell/cabal/pull/10104

**Comment by andreabedini** [comment]
@mpickering 
> I am talking about GHC's source distribution, for example - https://downloads.haskell.org/~ghc/9.10.1/ghc-9.10.1-src.tar.xz. In there you will find Lexer.hs.

:facepalm: Writing `build-tool-depends: alex` is accurate in Cabal-syntax source distribution. Since Hadrian already fiddles with it, Hadrian might as well remove the build-tool-depends line itself.

@geekosaur A more general solution that does not involve messing up with flags is to give an option to Setup.hs to ignore checking for the build-tool-depends; so redistributors can pre-process the souce distribuition at will and Cabal won't complain. I was under the impression this had been brought up already but I could not find any trace in the issues.

**Comment by mpickering** [comment]
@andreabedini 
Are you supporting adding a flag to `configure` which allows skipping the check for `build-tool-depends`? 

Are you willing to block #10104 from landing? 

Are you advocating that hadrian should be modified to patch Cabal-syntax in an ad-hoc manner rather than adding a flag?

**Comment by andreabedini** [comment]
> Are you supporting adding a flag to configure which allows skipping the check for build-tool-depends?

Yes. I generally speaking I believe it is a good thing to give users the flexibility to say "trust me". Checks are good for user experience but power-users (as ghc) need to be able to go past them. I am sure this was brought up before.

> Are you willing to block https://github.com/haskell/cabal/pull/10104 from landing?

Yes. *That one* would be an ad-hoc solution for an issue that can be solved in a better way.

> Are you advocating that hadrian should be modified to patch Cabal-syntax in an ad-hoc manner rather than adding a flag?

Everything that hadrian does is ad-hoc. It is a build system specifically written to build GHC. For me it has _carte blanche_. Deleting the build-tool-depends line from Cabal-syntax.cabal is as ad-hoc as adding Lexer.hs to the source distribution.

**Comment by mpickering** [comment]
Ok, we will implement the solutions as you advocate for. 

I will just comment that I think maintainers of downstream packages (Cabal) should be mindful of an important upstream when making changes to the packaging.

**Comment by andreabedini** [comment]
```
~/code/ghc master
λ git rev-parse HEAD
64fba310c2d23a41c88514aed0a482fbe5a3b184

~/code/ghc master
λ ./hadrian/build -j source-dist
Up to date
| Done
Build completed in 1.24s

~/code/ghc master
λ ls _build/source-dist/ghc-9.11.20240615-src.tar.xz
_build/source-dist/ghc-9.11.20240615-src.tar.xz

~/code/ghc master
λ cd /tmp/tmp.ku2CpcHBeF/

/tmp/tmp.ku2CpcHBeF
λ tar xJf ~/code/ghc/_build/source-dist/ghc-9.11.20240615-src.tar.xz

/tmp/tmp.ku2CpcHBeF
λ cd ghc-9.11.20240615

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ which alex
/usr/bin/which: no alex in (/home/andrea/.local/state/nix/profile/bin:/nix/var/nix/profiles/default/bin:/home/andrea/.cabal/bin:/home/andrea/.ghcup/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ ./configure
...
   windres      :
   genlib       :
   Happy        :  ()
   Alex         :  ()
   sphinx-build : /usr/bin/sphinx-build
   xelatex      : /usr/bin/xelatex
   makeinfo     : /usr/bin/makeinfo
   git          : /home/andrea/.local/state/nix/profile/bin/git
   cabal-install : /home/andrea/.ghcup/bin/cabal
...

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ echo 'extra-packages: alex' >> hadrian/cabal.project

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ hadrian/build -j
...
/----------------------------------------------------------\
| Successfully built program 'ghc-bin' (Stage1).           |
| Executable: _build/stage1/bin/ghc                        |
| Program synopsis: The Glorious Glasgow Haskell Compiler. |
\----------------------------------------------------------/
Build completed in 20m17s
```

Nothing has to change in cabal to get this to work.

**Comment by andreabedini** [comment]
Interestingly, the ghc-9.10.1 source distribution seems to behave differently. I haven't figured out why.

**Comment by juhp** [comment]
tl;dr but just noting for the record that Fedora ghc packages have build-depended on alex since 8.6 and also on happy since moving to Hadrian, ie from 9.2 on (there was also dependencies earlier times during ghc6...). You intrepret that how you will :-)

**Comment by andreabedini** [comment]
```
~/code/ghc master
λ git rev-parse HEAD
64fba310c2d23a41c88514aed0a482fbe5a3b184

~/code/ghc master
λ ./hadrian/build -j source-dist
Up to date
| Done
Build completed in 1.24s

~/code/ghc master
λ ls _build/source-dist/ghc-9.11.20240615-src.tar.xz
_build/source-dist/ghc-9.11.20240615-src.tar.xz

~/code/ghc master
λ cd /tmp/tmp.ku2CpcHBeF/

/tmp/tmp.ku2CpcHBeF
λ tar xJf ~/code/ghc/_build/source-dist/ghc-9.11.20240615-src.tar.xz

/tmp/tmp.ku2CpcHBeF
λ cd ghc-9.11.20240615

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ which alex
/usr/bin/which: no alex in (/home/andrea/.local/state/nix/profile/bin:/nix/var/nix/profiles/default/bin:/home/andrea/.cabal/bin:/home/andrea/.ghcup/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ ./configure
...
   windres      :
   genlib       :
   Happy        :  ()
   Alex         :  ()
   sphinx-build : /usr/bin/sphinx-build
   xelatex      : /usr/bin/xelatex
   makeinfo     : /usr/bin/makeinfo
   git          : /home/andrea/.local/state/nix/profile/bin/git
   cabal-install : /home/andrea/.ghcup/bin/cabal
...

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ echo 'extra-packages: alex' >> hadrian/cabal.project

/tmp/tmp.ku2CpcHBeF/ghc-9.11.20240615
λ hadrian/build -j

/----------------------------------------------------------\
| Successfully built program 'ghc-bin' (Stage1).           |
| Executable: _build/stage1/bin/ghc                        |
| Program synopsis: The Glorious Glasgow Haskell Compiler. |
\----------------------------------------------------------/
Build completed in 20m03s
...

**Comment by juhp** [comment]
> tl;dr but just noting for the record that Fedora ghc packages have build-depended on alex since 8.6 and also on happy since moving to Hadrian, ie from 9.2 on (there was also dependencies earlier times during ghc6...). You can intrepret that how you will :-)

Okay, I can correct myself: I got curious and at least Fedora ghc9.4, ghc9.6 and ghc9.8 can build without alex and happy: glad to learn this. ghc9.2 still seems to be needs alex though.
